FreeRTOS/Demo/WIN32-MSVC/main_blinky.c

```diff
@@ -237,37 +243,63 @@ uint32_t ulReceivedValue;
-		using console IO so it is ok.  However, note the comments at the top of
-		this file about the risks of making Windows system calls (such as
-		console output) from a FreeRTOS task. */
-		if( ulReceivedValue == mainVALUE_SENT_FROM_TASK )
-		{
-			printf( "Message received from task - idle time %llu%%\r\n", ulTaskGetIdleRunTimePercent() );
-		}
-		else if( ulReceivedValue == mainVALUE_SENT_FROM_TIMER )
-		{
-			printf( "Message received from software timer\r\n" );
-		}
-		else
-		{
-			printf( "Unexpected message\r\n" );
-		}
-
-		/* Reset the timer if a key has been pressed.  The timer will write
-		mainVALUE_SENT_FROM_TIMER to the queue when it expires. */
-		if( _kbhit() != 0 )
-		{
-			/* Remove the key from the input buffer. */
-			( void ) _getch();
-
-			/* Reset the software timer. */
-			xTimerReset( xTimer, portMAX_DELAY );
-		}
+        /* Enter critical section to use printf. Not doing this could potentially cause
+           a deadlock if the FreeRTOS simulator switches contexts and another task
+           tries to call printf - it should be noted that use of printf within
+           the FreeRTOS simulator is unsafe, but used here for simplicity. */
+        taskENTER_CRITICAL();
+        {
+            /*  To get here something must have been received from the queue, but
+            is it an expected value?  Normally calling printf() from a task is not
+            a good idea.  Here there is lots of stack space and only one task is
+            using console IO so it is ok.  However, note the comments at the top of
+            this file about the risks of making Windows system calls (such as
+            console output) from a FreeRTOS task. */
+            if (ulReceivedValue == mainVALUE_SENT_FROM_TASK)
+            {
+                printf("Message received from task - idle time %llu%%\r\n", ulTaskGetIdleRunTimePercent());
+            }
+            else if (ulReceivedValue == mainVALUE_SENT_FROM_TIMER)
+            {
+                printf("Message received from software timer\r\n");
+            }
+            else
+            {
+                printf("Unexpected message\r\n");
+            }
+        }
+        taskEXIT_CRITICAL();
 	}
 }
 /*-----------------------------------------------------------*/
 
+/* Called from prvKeyboardIntteruptSimulatorTask(), which is defined in main.c. */
+void vBlinkyKeyboardInterruptHandler( int xKeyPressed )
+{
+    /* Handle keyboard input. */
+    switch ( xKeyPressed )
+    {
+    case mainRESET_TIMER_KEY:
+
+        if ( xTimer != NULL )
+        {
+            /* Critical section around printf to prevent a deadlock
+               on context switch. */
+            taskENTER_CRITICAL();
+            {
+                printf("\r\nResetting software timer.\r\n\r\n");
+            }
+            taskEXIT_CRITICAL();
+
+            /* Reset the software timer. */
+            xTimerReset( xTimer, portMAX_DELAY );

```

Comment:
Is this intentionally left out of the task critical portion?

Reply:
hmm. it isn't necessary to include it in it. the reason the critical section is there in the first place is just to wrap the printf to prevent a context switch while calling it as that will cause deadlock if another task calls printf before this finishes. it wouldn't hurt it if it was but this is fine.